# Phase1:

## Goal of Collecting Dataset:

The aim of analyzing this football match dataset is to conduct a comprehensive performance evaluation of teams and players by examining key metrics such as goals, assists, and other relevant statistics. By leveraging historical data, we aim to build predictive models capable of forecasting match outcomes and individual player performances. Additionally, the analysis seeks to uncover trends and patterns in match results, player statistics, and team strategies, which can provide valuable insights into game dynamics and competitive behaviors.

To achieve these objectives, we will employ clustering and classification techniques using machine learning methods learned during the course.

## Source of Dataset: 
Kaggle Platform Dataset link: https://www.kaggle.com/datasets/evangower/fifa-world-cup

## General Information about the Dataset:

Number of Attributes:15

Number of Objects:900 (Match)

Class Name/Labels: outcome


In [13]:

import pandas as pd

df = pd.read_csv('Dataset/wcmatches.csv')

print(df.info())

num_row = len(df)
print("Number of rows:", num_row)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 900 entries, 0 to 899
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   year            900 non-null    int64 
 1   country         900 non-null    object
 2   city            900 non-null    object
 3   stage           900 non-null    object
 4   home_team       900 non-null    object
 5   away_team       900 non-null    object
 6   home_score      900 non-null    int64 
 7   away_score      900 non-null    int64 
 8   outcome         900 non-null    object
 9   win_conditions  62 non-null     object
 10  winning_team    731 non-null    object
 11  losing_team     731 non-null    object
 12  date            900 non-null    object
 13  month           900 non-null    object
 14  dayofweek       900 non-null    object
dtypes: int64(3), object(12)
memory usage: 105.6+ KB
None
Number of rows: 900


------------------------------------------------------

# Phase2:

| Attributes Name | Data type | Description | Possible Values |
|----------|----------|----------|----------|
|Year |	Integer| The year of the match |	Any year (e.g., 1930, 1934, etc.)|
|Country |	String |The country hosting the World Cup |  Categorical string values (different possible)|
|City	| String |	The city where the match took place	 | Categorical string values (different possible)|
|Stage	| String|	The stage of the tournament | Categorical string values (different possible)|
|Home Team |	String | The team playing at home |	Categorical string values (different possible)|
|Away Team |	String |	The team playing away |	Categorical string values (different possible)|
|Home Score |	Integer	| The score of the home team |	0-10|
|Away Score	| Integer |	The score of the away team |	0-8|
|Outcome |	String | The result of the match |	H (Home win), A (Away win), D (Draw)|
|Win Conditions |	String |	Conditions under which the winning team was determined	Specific phrases |Categorical string values (different possible)|
|Winning Team |	String | The team that won the match |	Categorical string values (different possible)|
|Losing Team|	String | The team that lost the match |	Categorical string values (different possible)|
|Date |Date |	The date of the match |	Datetime |
|Month |String	|The month in which the match took place|Categorical string values (different possible)|
|DayOfWeek |	String |	The day of the week when the match occurred | Categorical string values (different possible)|


## Sample

In [14]:

sample = df.sample(frac=1)
print(sample);

     year        country              city          stage    home_team  \
84   1954    Switzerland          Lausanne        Group 1       Brazil   
37   1938         France          Toulouse    Round of 16         Cuba   
592  2002          Japan           Saitama        Group H        Japan   
194  1966        England        Sunderland  Quarterfinals      Hungary   
374  1986         Mexico       Mexico City        Group A  South Korea   
..    ...            ...               ...            ...          ...   
859  2018         Russia  Saint Petersburg        Group E       Brazil   
1    1930        Uruguay        Montevideo        Group 4      Belgium   
899  2018         Russia            Moscow          Final       France   
450  1990          Italy             Turin    Round of 16       Brazil   
469  1994  United States           Orlando        Group F      Belgium   

         away_team  home_score  away_score outcome win_conditions  \
84      Yugoslavia           1           1

## Show the Min., 1st Qu., Median, Mean ,3rd Qu.,Max. for each numeric column:

In [16]:
summary_stats=df.describe();
print(summary_stats);

              year  home_score  away_score
count   900.000000  900.000000  900.000000
mean   1986.915556    1.568889    1.262222
std      23.150270    1.489574    1.313960
min    1930.000000    0.000000    0.000000
25%    1970.000000    0.000000    0.000000
50%    1990.000000    1.000000    1.000000
75%    2006.000000    2.000000    2.000000
max    2018.000000   10.000000    8.000000


- Year: The range of year is between 1930 and 2018, with an average year of 1987. Most matches occurred between 1970 and 2006, with a median year of
1990, indicating a relatively even distribution of matches over time.

- Home Score: Home team scores range from 0 to 10 goals, with an average of 1.57 goals per match. While 25% of the matches saw the home team
scoring no goals, the median score is 1, and in 75% of the matches, the home team scored 2 or fewer goals.

- Away Score: Away team scores range from 0 to 8 goals, with an average of 1.26 goals per match. Similar to home teams, 25% of the matches had away
teams scoring no goals, and the median score is 1, with 75% of the away teams scoring 2 or fewer goals.

## Show the Variance:

In [17]:
variance= df.var(numeric_only=True);
print(variance);

year          535.935019
home_score      2.218831
away_score      1.726491
dtype: float64


Show the Variance :
Variance shows how much values spread out. High variance means more spread, while low variance means values are closer to the average. So, our
variance results show:


Year: The high variance shows significant spread in match years, which is expected due to the wide range [1930-2018]

home score:The variance is relatively low which means there isn't much variation in the scores of the home teams

Away_score: The variance is also low, indicating that there is little variation in the scores of the away teams

## Graphs:

### Understanding the data through graph representations:

To understand the relationship between match outcomes and all available attributes, particularly how they are associated with predicting match results and team performance, the focus was primarily on variables related to teams and final results. This analysis identifies whether a team won or lost, utilizing all attributes to extract relationships and infer whether an increase in a specific attribute (such as goals scored or possession) suggests a higher likelihood of winning the match.

Additionally, the analysis helps determine whether performance is more influenced by factors like venue (home or away) or the ranking differences between teams. This aids in identifying key factors that impact match outcomes and discovering indicators that could assist in predicting future wins or losses.

///

## Data Preprocessing:

### Data Cleaning:

#### 1-Remove duplicates:

##### First, we count the number of duplicates rows, and then we remove them:

In [18]:
import pandas as pd
data = pd.read_csv('Dataset/wcmatches.csv')
num_duplicates = data.duplicated().sum()
data_cleaned = data.drop_duplicates()
print("Number of duplicate rows:", num_duplicates)
print("DataFrame after dropping all duplicate rows:")
print(data_cleaned)
data_cleaned.to_csv('Cleaned_dataset.csv', index=False)

Number of duplicate rows: 0
DataFrame after dropping all duplicate rows:
     year  country              city          stage  home_team      away_team  \
0    1930  Uruguay        Montevideo        Group 1     France         Mexico   
1    1930  Uruguay        Montevideo        Group 4    Belgium  United States   
2    1930  Uruguay        Montevideo        Group 2     Brazil     Yugoslavia   
3    1930  Uruguay        Montevideo        Group 3       Peru        Romania   
4    1930  Uruguay        Montevideo        Group 1  Argentina         France   
..    ...      ...               ...            ...        ...            ...   
895  2018   Russia             Sochi  Quarterfinals     Russia        Croatia   
896  2018   Russia  Saint Petersburg     Semifinals     France        Belgium   
897  2018   Russia            Moscow     Semifinals    Croatia        England   
898  2018   Russia  Saint Petersburg    Third place    Belgium        England   
899  2018   Russia            Moscow

This shows that our dataset has no duplicate rows, so the total number of rows remains unchanged at 900.

#### 2-Handling Missing Values: 

##### First, we calculate the number of missing values in our dataset:

In [19]:
import pandas as pd
data = pd.read_csv('Cleaned_dataset.csv');
sample=data.sample(n=20);
print(sample);
missing_values = data.isna()
missing_counts = missing_values.sum();
rows_with_missing = data[data.isna().any(axis=1)];

print("Missing values in each column:");
print(missing_counts);
print("\nRows with missing values:");
print(rows_with_missing);

     year       country            city          stage         home_team  \
229  1970        Mexico     Mexico City     Semifinals      West Germany   
650  2006       Germany       Nuremberg        Group D            Mexico   
430  1990         Italy           Turin        Group C            Brazil   
96   1954   Switzerland          Geneva  Quarterfinals      West Germany   
5    1930       Uruguay      Montevideo        Group 1             Chile   
848  2018        Russia           Sochi        Group G           Belgium   
460  1990         Italy          Naples     Semifinals             Italy   
381  1986        Mexico  Nezahualcóyotl        Group E           Denmark   
771  2010  South Africa    Johannesburg          Final       Netherlands   
177  1966       England   Middlesbrough        Group 4             Chile   
799  2014        Brazil       Fortaleza        Group G           Germany   
397  1986        Mexico            León    Round of 16      Soviet Union   
282  1978   

This shows that our dataset has missing values in multiple columns: 838 in the 'win_conditions' column, 169 in the 'winning_team' column, and 169 in the 'losing_team' column.

##### Secondly, we handle these missing values by calculating the mode value for the target columns ['win_conditions', 'winning_team', 'losing_team'], and then we replace the missing values with the mode.

In [20]:
import pandas as pd

data = pd.read_csv('Cleaned_dataset.csv')

data.columns = data.columns.str.strip()


for column in ['win_conditions', 'winning_team', 'losing_team']:
    modal_value = data[column].mode()[0]  
    data[column] = data[column].fillna(value=modal_value)  

print("Missing values after filling:")
print(data.isnull().sum())

data.to_csv('Cleaned_dataset.csv', index=False)



Missing values after filling:
year              0
country           0
city              0
stage             0
home_team         0
away_team         0
home_score        0
away_score        0
outcome           0
win_conditions    0
winning_team      0
losing_team       0
date              0
month             0
dayofweek         0
dtype: int64


and Here, the output shows that we handled all the missing values.

#### 3-Handling Outliers:

##### Before we handle our outliers we did choose the IQR method Based on our dataset to identify how many outliers in our dataset.

In [21]:
import pandas as pd
data = pd.read_csv('Cleaned_dataset.csv')
import numpy as np
outlier_threshold = 1.5

def count_outliers(column_data):
    q1 = np.percentile(column_data, 25)
    q3 = np.percentile(column_data, 75)
    iqr = q3 - q1
    upper_bound = q3 + outlier_threshold * iqr
    lower_bound = q1 - outlier_threshold * iqr
    outliers = (column_data > upper_bound) | (column_data < lower_bound)
    return sum(outliers)

numeric_columns = data.select_dtypes(include=[np.number]).columns

outlier_counts = {}
total_rows_with_outliers = 0

for column in numeric_columns:
    outliers = count_outliers(data[column])
    outlier_counts[column] = outliers
    total_rows_with_outliers += outliers

total_rows = len(data)

print("Outlier Counts:")
for column, count in outlier_counts.items():
    print(f"{column}: {count} rows with outliers")

print(f"Total Rows with Outliers: {total_rows_with_outliers}")

Outlier Counts:
year: 0 rows with outliers
home_score: 24 rows with outliers
away_score: 11 rows with outliers
Total Rows with Outliers: 35


The analysis of our dataset revealed that a total of 35 rows contain outliers, which is a small proportion compared to the overall dataset of 900 rows. Specifically, there are 0 rows with outliers in the 'year' column, 24 rows in the 'home_score' column, and 11 rows in the 'away_score' column. Although this represents a limited number of outliers, their presence indicates some extreme values that deviate from the majority of the data. To address this issue, we have decided to adopt a strategy of capping the outliers. Rather than removing these data points entirely, we opted to replace them with the nearest non-outlier values. This approach allows us to retain the entirety of our dataset while mitigating the impact of these extreme values on our subsequent analyses. By capping the outliers and maintaining the remaining data, we aim to preserve the integrity of the dataset and avoid the potential loss of valuable information that would occur through outright removal of the outliers. This ensures that our analysis incorporates as much relevant information as possible while still accounting for the presence of extreme observations

##### Handling Outliers:

In [22]:
import pandas as pd
import numpy as np

data = pd.read_csv('Cleaned_dataset.csv')

outlier_threshold = 1.5

def count_outliers(column_data):
    q1 = np.percentile(column_data, 25)
    q3 = np.percentile(column_data, 75)
    iqr = q3 - q1
    upper_bound = q3 + outlier_threshold * iqr
    lower_bound = q1 - outlier_threshold * iqr
    outliers = (column_data > upper_bound) | (column_data < lower_bound)
    return sum(outliers)

numeric_columns = data.select_dtypes(include=[np.number]).columns

outlier_counts = {}
total_rows = len(data)

for column in numeric_columns:
    outliers = count_outliers(data[column])
    outlier_counts[column] = outliers
    
    q1 = np.percentile(data[column], 25)
    q3 = np.percentile(data[column], 75)
    iqr = q3 - q1
    upper_bound = q3 + outlier_threshold * iqr
    lower_bound = q1 - outlier_threshold * iqr
    data[column] = np.clip(data[column], lower_bound, upper_bound)

data.to_csv('Cleaned_dataset.csv', index=False)


##### Checking the results by counting outliers after handling them.

In [23]:
import pandas as pd
import numpy as np

data1 = pd.read_csv('Cleaned_dataset.csv')

outlier_threshold = 1.5

def count_outliers(column_data1):
    q1 = np.percentile(column_data1, 25)
    q3 = np.percentile(column_data1, 75)
    iqr = q3 - q1
    upper_bound = q3 + outlier_threshold * iqr
    lower_bound = q1 - outlier_threshold * iqr
    outliers = (column_data1 > upper_bound) | (column_data1 < lower_bound)
    return sum(outliers)


numeric_columns = data1.select_dtypes(include=[np.number]).columns


outlier_counts = {}
total_rows_with_outliers = 0


for column in numeric_columns:
    outliers = count_outliers(data1[column])
    outlier_counts[column] = outliers
    total_rows_with_outliers += outliers

total_rows = len(data1)

print("Outlier Counts:")
for column, count in outlier_counts.items():
    print(f"{column}: {count} rows with outliers")

print(f"Total Rows with Outliers: {total_rows_with_outliers}")


Outlier Counts:
year: 0 rows with outliers
home_score: 0 rows with outliers
away_score: 0 rows with outliers
Total Rows with Outliers: 0


### Data Transmission

#### 1-Encoding

In [1]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder

data1 = pd.read_csv('Cleaned_dataset.csv')

le = LabelEncoder()
data1['country'] = le.fit_transform(data1['country'])
data1['city'] = le.fit_transform(data1['city'])
data1['stage'] = le.fit_transform(data1['stage'])
data1['dayofweek'] = le.fit_transform(data1['dayofweek'])
data1['month'] = le.fit_transform(data1['month'])
data1['home_team'] = le.fit_transform(data1['home_team'])
data1['away_team'] = le.fit_transform(data1['away_team'])
data1['losing_team'] = le.fit_transform(data1['losing_team'])
data1['winning_team'] = le.fit_transform(data1['winning_team'])

print('Encoded DataFrame')
print(data1.head())

Encoded DataFrame
   year  country  city  stage  home_team  away_team  home_score  away_score  \
0  1930       16    83      3         27         42           4           1   
1  1930       16    83      6          5         76           0           3   
2  1930       16    83      4          8         80           1           2   
3  1930       16    83      5         52         56           1           3   
4  1930       16    83      3          2         27           1           0   

  outcome    win_conditions  winning_team  losing_team        date  month  \
0       H  Italy won in AET            21           45  1930-07-13      0   
1       A  Italy won in AET            61            5  1930-07-13      0   
2       A  Italy won in AET            65            8  1930-07-14      0   
3       A  Italy won in AET            44           55  1930-07-14      0   
4       H  Italy won in AET             1           29  1930-07-15      0   

   dayofweek  
0          3  
1          3  

several nominal attributes have been encoded using the LabelEncoder. These attributes include country, city, stage, dayofweek, and month, home_team, away_team, losing_team, and winning_team. Each of these columns has been transformed from its original categorical form into a numerical format. For instances cities being numbered '83' to perform analysis

#### 2-Normalization

In [3]:
import pandas as pd
data1 = pd.read_csv('Cleaned_dataset.csv')

columns_to_normalize = ['home_score', 'away_score']

for column in columns_to_normalize:
    max_abs_value = data1[column].abs().max()
    data1[column] = data1[column] / (10 ** len(str(int(max_abs_value))))


print("DataFrame after Decimal Scaling Normalization:")
print(data1.head())

DataFrame after Decimal Scaling Normalization:
   year  country        city    stage  home_team      away_team  home_score  \
0  1930  Uruguay  Montevideo  Group 1     France         Mexico         0.4   
1  1930  Uruguay  Montevideo  Group 4    Belgium  United States         0.0   
2  1930  Uruguay  Montevideo  Group 2     Brazil     Yugoslavia         0.1   
3  1930  Uruguay  Montevideo  Group 3       Peru        Romania         0.1   
4  1930  Uruguay  Montevideo  Group 1  Argentina         France         0.1   

   away_score outcome    win_conditions   winning_team losing_team  \
0         0.1       H  Italy won in AET         France      Mexico   
1         0.3       A  Italy won in AET  United States     Belgium   
2         0.2       A  Italy won in AET     Yugoslavia      Brazil   
3         0.3       A  Italy won in AET        Romania        Peru   
4         0.0       H  Italy won in AET      Argentina      France   

         date month dayofweek  
0  1930-07-13   Jul    Su

Home and away scored have been normalized using decimal scaling to  become values between [0,1]

#### 3-Aggregation

In [7]:
import pandas as pd
data1 = pd.read_csv('Cleaned_dataset.csv')

country_stats = data1.groupby('country').agg(
    home_score_mean=('home_score', 'mean'),
    away_score_mean=('away_score', 'mean'),
    total_wins=('winning_team', lambda x: (x == x.mode()[0]).sum())  
).reset_index()

print("Country Stats:")
print(country_stats)

Country Stats:
          country  home_score_mean  away_score_mean  total_wins
0       Argentina         1.421053         1.210526          13
1          Brazil         1.441860         1.476744          20
2           Chile         1.406250         1.343750          10
3         England         1.437500         1.343750           6
4          France         1.756098         1.280488          27
5         Germany         1.382353         0.941176          28
6           Italy         1.478261         1.173913          12
7           Japan         1.125000         1.000000          12
8          Mexico         1.476190         1.202381          26
9          Russia         1.359375         1.265625          12
10   South Africa         1.171875         1.062500          17
11    South Korea         1.625000         1.187500           9
12          Spain         1.730769         0.980769          20
13         Sweden         1.914286         1.571429          15
14    Switzerland        

The aggregation summarizes match data by country, calculating the average home and away team scores, as well as counting the number of wins for the most frequent winning team in each country.

#### 4-Discretization

In [9]:
import pandas as pd
data1 = pd.read_csv('Cleaned_dataset.csv')

year_labels = ["1930s", "1940s", "1950s", "1960s", "1970s", "1980s", "1990s", "2000s", "2010s",]
year_bins = [1929, 1939, 1949, 1959, 1969, 1979, 1989, 1999, 2009, 2019,]

data1["Year_Decade"] = pd.cut(data1["year"], bins=year_bins, labels=year_labels, right=False)
data1['date'] = pd.to_datetime(data1['date'])
data1['Month'] = data1['date'].dt.month_name()

print(data1.head())

   year  country        city    stage  home_team      away_team  home_score  \
0  1930  Uruguay  Montevideo  Group 1     France         Mexico           4   
1  1930  Uruguay  Montevideo  Group 4    Belgium  United States           0   
2  1930  Uruguay  Montevideo  Group 2     Brazil     Yugoslavia           1   
3  1930  Uruguay  Montevideo  Group 3       Peru        Romania           1   
4  1930  Uruguay  Montevideo  Group 1  Argentina         France           1   

   away_score outcome    win_conditions   winning_team losing_team       date  \
0           1       H  Italy won in AET         France      Mexico 1930-07-13   
1           3       A  Italy won in AET  United States     Belgium 1930-07-13   
2           2       A  Italy won in AET     Yugoslavia      Brazil 1930-07-14   
3           3       A  Italy won in AET        Romania        Peru 1930-07-14   
4           0       H  Italy won in AET      Argentina      France 1930-07-15   

  month dayofweek Year_Decade Month  


discretizes the "year" column into decades by categorizing it into decade ranges, assigns each entry a decade label, converts the "date" column to a datetime format, and extracts the month name from the date into a new "Month" column

## Feature Selection:

- Correlation Coefficient:

- Chi squared:

- Drop Out Highly Correlated Attributes: